aptos-move/framework/aptos-framework/sources/jwks.move

```diff
@@ -0,0 +1,701 @@
+    /// Observing `UnsupportedJWK`s means the providers adopted a new key type/format, and the system should be updated.
+    struct UnsupportedJWK has copy, drop, store {
+        id: vector<u8>,
+        payload: vector<u8>,
+    }
+
+    /// A JWK variant where `kty` is `RSA`.
+    struct RSA_JWK has copy, drop, store {
+        kid: String,
+        kty: String,
+        alg: String,
+        e: String,
+        n: String,
+    }
+
+    /// A JSON web key.
+    struct JWK has copy, drop, store {
+        /// A `JWK` variant packed as an `Any`.
+        /// Currently the variant type is one of the following.
+        /// - `RSA_JWK`
+        /// - `UnsupportedJWK`
+        variant: Any,
+    }
+
+    /// A provider and its `JWK`s.
+    struct ProviderJWKs has copy, drop, store {
+        /// The utf-8 encoding of the issuer string (e.g., "https://www.facebook.com").
+        issuer: vector<u8>,
+
+        /// Bumped every time the JWKs for the current issuer is updated.
+        version: u64,
+
+        /// The `jwks` each has a unique `id` and are sorted by `id` in dictionary order.
+        jwks: vector<JWK>,
+    }
+
+    /// Multiple `ProviderJWKs`s, indexed by issuer and key ID.
+    struct JWKs has copy, drop, store {
+        /// Entries each has a unique `issuer`, and are sorted by `issuer` in dictionary order.
+        entries: vector<ProviderJWKs>,
+    }
+
+    /// The `JWKs` that validators observed and agreed on.
+    struct ObservedJWKs has copy, drop, key, store {
+        jwks: JWKs,
+    }
+
+    #[event]
+    /// When the `ObservedJWKs` is updated, this event is sent to resync the JWK consensus state in all validators.
+    struct ObservedJWKsUpdated has drop, store {
+        epoch: u64,
+        jwks: JWKs,
+    }
+
+    /// A small edit that can be applied to a `JWKs`.
+    struct JWKPatch has copy, drop, store {
+        /// A `JWKPatch` variant packed as an `Any`.
+        /// Currently the variant type is one of the following.
+        /// - `JWKPatchRemoveAll`
+        /// - `JWKPatchRemoveIssuer`
+        /// - `JWKPatchRemoveJWK`
+        /// - `JWKPatchUpsertJWK`
+        variant: Any,
+    }
+
+    /// A `JWKPatch` variant to remove all JWKs.
+    struct JWKPatchRemoveAll has copy, drop, store {}
+
+    /// A `JWKPatch` variant to remove all JWKs from an issuer.
+    struct JWKPatchRemoveIssuer has copy, drop, store {
+        issuer: vector<u8>,
+    }
+
+    /// A `JWKPatch` variant to remove a JWK.
+    struct JWKPatchRemoveJWK has copy, drop, store {
+        issuer: vector<u8>,
+        jwk_id: vector<u8>,
+    }
+
+    /// A `JWKPatch` variant to upsert a JWK.
+    struct JWKPatchUpsertJWK has copy, drop, store {
+        issuer: vector<u8>,
+        jwk: JWK,
+    }
+
+    /// A sequence of `JWKPatch` that needs to be applied *one by one* to the `ObservedJWKs`.
+    ///
+    /// Maintained by governance proposals.
+    struct JWKPatches has key {
```

Comment:
Do validators agree on the JWKPatches or the FinalJWKs?

Reply:
They agree on `ObservedJWKs` and do not care `JWKPatches`. FinalJWKs is always the derivation of `ObservedJWKs + JWKPatches`.

Reply:
Where do we use (produce/consume) JWKPatches?

Reply:
See function `regenerate_final_jwks` [here](https://github.com/aptos-labs/aptos-core/pull/11372/files/dfbd1c988ec64163079906dbe1d0f88c6f50fcca#diff-4e80bb4713479b715df7b82d4c9b032b8e4a4598a1242884497c7245520e8bc9R290).

Whenever ObservedJWKs/JWKPatches is updated, we call it.

Reply:
and JWKPatches is only updated by governance proposals.